batch/test/test_dag.py

```diff
@@ -0,0 +1,204 @@
+        head = batch.create_job('alpine:3.8', command=['echo', 'head'])
+        head_id = head.id
+        head.delete()
+        tail = batch.create_job('alpine:3.8', command=['echo', 'tail'], parents=[head_id])
+    except requests.exceptions.HTTPError as err:
+        assert err.response.status_code == 400
+        assert re.search('.*invalid parent: no job with id.*', err.response.text)
+        return
+    assert False
+
+
+def test_dag(client):
+    batch = client.create_batch()
+    head = batch.create_job('alpine:3.8', command=['echo', 'head'])
+    left = batch.create_job('alpine:3.8', command=['echo', 'left'], parents=[head.id])
+    right = batch.create_job('alpine:3.8', command=['echo', 'right'], parents=[head.id])
+    tail = batch.create_job('alpine:3.8', command=['echo', 'tail'], parents=[left.id, right.id])
+    status = batch.wait()
+    assert status['jobs']['Complete'] == 4
+    for node in [head, left, right, tail]:
+        status = node.status()
+        assert status['state'] == 'Complete'
+        assert status['exit_code'] == 0
+
+
+def test_cancel_tail(client):
+    batch = client.create_batch()
+    head = batch.create_job('alpine:3.8', command=['echo', 'head'])
+    left = batch.create_job('alpine:3.8', command=['echo', 'left'], parents=[head.id])
+    right = batch.create_job('alpine:3.8', command=['echo', 'right'], parents=[head.id])
+    tail = batch.create_job('alpine:3.8', command=['echo', 'tail'], parents=[left.id, right.id])
+    tail.cancel()
+    status = batch.wait()
+    assert status['jobs']['Complete'] == 3
+    for node in [head, left, right]:
+        status = node.status()
+        assert status['state'] == 'Complete'
+        assert status['exit_code'] == 0
+    assert tail.status()['state'] == 'Cancelled'
+
+
+def test_cancel_left_before_tail(client):
+    batch = client.create_batch()
+    head = batch.create_job('alpine:3.8', command=['echo', 'head'])
+    left = batch.create_job('alpine:3.8', command=['echo', 'left'], parents=[head.id])
+    left.cancel()
+    right = batch.create_job('alpine:3.8', command=['echo', 'right'], parents=[head.id])
+    tail = batch.create_job('alpine:3.8', command=['echo', 'tail'], parents=[left.id, right.id])
+    status = batch.wait()
+    assert status['jobs']['Complete'] == 2
+    for node in [head, right]:
+        status = node.status()
+        assert status['state'] == 'Complete'
+        assert status['exit_code'] == 0
+    assert left.status()['state'] == 'Cancelled'
+    assert tail.status()['state'] == 'Cancelled'
+
+
+def test_cancel_left_after_tail(client):
+    batch = client.create_batch()
+    head = batch.create_job('alpine:3.8', command=['echo', 'head'])
+    left = batch.create_job('alpine:3.8', command=['echo', 'left'], parents=[head.id])
+    right = batch.create_job('alpine:3.8', command=['echo', 'right'], parents=[head.id])
+    tail = batch.create_job('alpine:3.8', command=['echo', 'tail'], parents=[left.id, right.id])
+    left.cancel()
```

Comment:
This seems like a race condition waiting to happen, all the jobs could complete before executing the cancel.

Reply:
addressed